Add kubelet and nodeExporter pushprox

dbason opened the pull request:
Fixes https://github.com/rancher/rancher/issues/32841

Adds a pushprox for the kubelet and for node-exporter which runs on the host network.

Kubelet ServiceMonitor is still enabled by default, will disable it using values in the dashboard after this is merged. The node-exporter ServiceMonitor is disabled, and the node-exporter pushprox is enabled by default.

This supercedes #1055

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/values.yaml.patch:
```diff
@@ -230,6 +243,35 @@
 +      - effect: "NoSchedule"
 +        operator: "Exists"
 +
++## hardened components PushProx Monitoring
```
Can we change this to `## Additional PushProx Monitoring`?

aiyengar2 commented on packages/rancher-monitoring/generated-changes/overlay/Chart.yaml.orig:
```diff
@@ -0,0 +1,101 @@
+annotations:
```
Chart.yaml.orig seems incorrect. This looks like a patch issue.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/Chart.yaml.patch:
```diff
@@ -16,7 +16,16 @@
+   name: hardenedKubelet
+   repository: file://./charts/hardenedKubelet
+-- condition: hardenedNodeExporter.enabled
++- condition: nodeExporter.enabled
```
I don't think we should change this condition from `hardenedNodeExporter.enabled` to `nodeExporter.enabled`

dbason replied:
Sure I can put it back.  The default value for hardenedNodeExporter is true; my thinking was to always enable the node exporter pushprox when the node exporter is enabled.  I can change that logic if you would prefer to have them separate.

aiyengar2 replied:
Let's keep them separate since the node-exporter flag is reserved for someone who wants to stick with the old approach (no additional pushproxes). This will be an opt-in flag to support deploying node-exporter in hardened clusters where even the nodeExporter port cannot be scraped.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/templates/exporters/node-exporter/servicemonitor.yaml.patch:
```diff
@@ -0,0 +1,8 @@
+--- charts-original/templates/exporters/node-exporter/servicemonitor.yaml
++++ charts/templates/exporters/node-exporter/servicemonitor.yaml
+@@ -1,4 +1,4 @@
+-{{- if .Values.nodeExporter.enabled }}
++{{- if (and .Values.nodeExporter.enabled .Values.nodeExporter.serviceMonitor.enabled) }}
```
This should be disabled if PushProx is enabled since PushProx already comes in with a ServiceMonitor

dbason commented:
@aiyengar2 resolved the minor stuff.  When I was looking through the values.yaml I also found a couple of duplicate entries which I have cleaned up.

Just waiting on confirmation for the nodeExporter settinngs.

dbason commented:
@aiyengar2 I think I've covered everything off

- hardenedKubelet, needed to make some RBAC additions for this to work as the kubelet port is authenticated.
- k3sServer: the 10249 port is exposed on agents as well, but unfortunately it doesn't include the cadvisor and probes endpoints that we need for kubelet monitoring, so I have switched to 10250.  Because the kubelet monitoring includes metric path labels this caused inflated numbers for some of the controlplane dashboard counts, so I've adjusted those with a helper.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/charts/hardenedKubelet/templates/pushprox-servicemonitor.yaml.patch:
```diff
@@ -0,0 +1,54 @@
++    relabelings:
++{{ toYaml .Values.relabelings | indent 4 }}
++{{- end }}
++{{- if .Values.cAdvisor }}
++  - port: metrics
++    proxyUrl: {{ template "pushProxy.proxyUrl" . }}
++    {{- if .Values.clients.https.enabled }}
++    params:
++      _scheme: [https]
++    {{- end }}
++    path: /metrics/cadvisor
++    honorLabels: true
++{{- if .Values.metricRelabelings }}
++    metricRelabelings:
++{{ tpl (toYaml .Values.metricRelabelings | indent 4) . }}
++{{- end }}
++{{- if .Values.relabelings }}
++    relabelings:
++{{ toYaml .Values.relabelings | indent 4 }}
++{{- end }}
++{{- end }}
++{{- if .Values.probes }}
++  - port: metrics
++    proxyUrl: {{ template "pushProxy.proxyUrl" . }}
++    {{- if .Values.clients.https.enabled }}
++    params:
++      _scheme: [https]
++    {{- end }}
++    path: /metrics/probes
++    honorLabels: true
++{{- if .Values.metricRelabelings }}
++    metricRelabelings:
++{{ tpl (toYaml .Values.metricRelabelings | indent 4) . }}
++{{- end }}
++{{- if .Values.relabelings }}
++    relabelings:
++{{ toYaml .Values.relabelings | indent 4 }}
++{{- end }}
++{{- end }}
```
Instead of patching pushprox specifically for hardenedKubelet, could we just update the rancher-pushprox chart to have `.Values.serviceMonitor.metricsConfig` and insert `{{ tpl (toYaml .Values.serviceMonitor.additionalEndpoints) | indent 2 }}` at the bottom of:

https://github.com/rancher/charts/blob/7836aff0c1b2846c3fbcf8ed2adc6963d21c20b1/charts/rancher-pushprox/rancher-pushprox/0.1.300/templates/pushprox-servicemonitor.yaml#L10-L15

We can then add this configuration to the values.yaml. This should come after `{{ tpl (toYaml .Values.serviceMonitor.additionalMetricsEndpointConfig) | indent 4 }}` from above.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/charts/hardenedKubelet/templates/pushprox-servicemonitor.yaml.patch:
```diff
@@ -0,0 +1,54 @@
+     params:
+       _scheme: [https]
+     {{- end }}
++    honorLabels: true
++{{- if .Values.metricRelabelings }}
++    metricRelabelings:
++{{ tpl (toYaml .Values.metricRelabelings | indent 4) . }}
++{{- end }}
++{{- if .Values.relabelings }}
++    relabelings:
++{{ toYaml .Values.relabelings | indent 4 }}
++{{- end }}
```
Can these values be added under `.Values.serviceMonitor.metricsEndpoint.*` instead of at the root level?

Also, let's add this to the original pushprox chart instead of patching it.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/charts/k3sServer/templates/pushprox-clients-rbac.yaml.patch:
```diff
@@ -0,0 +1,16 @@
+   - {{ template "pushProxy.client.name" . }}
+ {{- if and .Values.clients.https.enabled .Values.clients.https.useServiceAccountCredentials }}
+-- nonResourceURLs: ["/metrics"]
++- nonResourceURLs:
++  - "/metrics"
++  - "/metrics/cadvisor"
++  verbs: ["get"]
++- apiGroups: [""]
++  resources: ["nodes/metrics"]
```
same comments as above; let's avoid patching this subchart if it could be a commonly used feature anyways.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/charts/k3sServer/templates/pushprox-servicemonitor.yaml.patch:
```diff
@@ -0,0 +1,54 @@
+     params:
+       _scheme: [https]
+     {{- end }}
++    honorLabels: true
++{{- if .Values.metricRelabelings }}
++    metricRelabelings:
++{{ tpl (toYaml .Values.metricRelabelings | indent 4) . }}
++{{- end }}
++{{- if .Values.relabelings }}
++    relabelings:
++{{ toYaml .Values.relabelings | indent 4 }}
++{{- end }}
```
same comments as above; let's avoid patching this subchart if it could be a commonly used feature anyways.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/charts/k3sServer/templates/pushprox-servicemonitor.yaml.patch:
```diff
@@ -0,0 +1,54 @@
++    relabelings:
++{{ toYaml .Values.relabelings | indent 4 }}
++{{- end }}
++{{- if .Values.cAdvisor }}
++  - port: metrics
++    proxyUrl: {{ template "pushProxy.proxyUrl" . }}
++    {{- if .Values.clients.https.enabled }}
++    params:
++      _scheme: [https]
++    {{- end }}
++    path: /metrics/cadvisor
++    honorLabels: true
++{{- if .Values.metricRelabelings }}
++    metricRelabelings:
++{{ tpl (toYaml .Values.metricRelabelings | indent 4) . }}
++{{- end }}
++{{- if .Values.relabelings }}
++    relabelings:
++{{ toYaml .Values.relabelings | indent 4 }}
++{{- end }}
++{{- end }}
++{{- if .Values.probes }}
++  - port: metrics
++    proxyUrl: {{ template "pushProxy.proxyUrl" . }}
++    {{- if .Values.clients.https.enabled }}
++    params:
++      _scheme: [https]
++    {{- end }}
++    path: /metrics/probes
++    honorLabels: true
++{{- if .Values.metricRelabelings }}
++    metricRelabelings:
++{{ tpl (toYaml .Values.metricRelabelings | indent 4) . }}
++{{- end }}
++{{- if .Values.relabelings }}
++    relabelings:
++{{ toYaml .Values.relabelings | indent 4 }}
++{{- end }}
++{{- end }}
```
same comments as above; let's avoid patching this subchart if it could be a commonly used feature anyways.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/templates/_helpers.tpl.patch:
```diff
@@ -73,6 +91,20 @@
++{{- end -}}
++{{- end }}
++
++{{- define "exporter.controlplane.upLabels" -}}
++{{- if .Values.k3sServer.enabled -}}
++, metrics_path=\"/metrics\"
++{{- end -}}
++{{- end }}
++
```
I see why you added this, but I think this should not be a helper function. We should patch the expression directly for this, since it only happens a couple of times.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/templates/exporters/node-exporter/servicemonitor.yaml.patch:
```diff
@@ -0,0 +1,8 @@
++++ charts/templates/exporters/node-exporter/servicemonitor.yaml
+@@ -1,4 +1,4 @@
+-{{- if .Values.nodeExporter.enabled }}
++{{- if (and .Values.nodeExporter.enabled (not .Values.hardenedNodeExporter.enabled)) }}
```
Instead of this, let's fail if the user has both enabled.
```
{{- if .Values.nodeExporter.enabled }} 
{{- if .Values.hardenedNodeExporter.enabled }}
{{- required "Cannot set both .Values.nodeExporter.enabled and .Values.hardenedNodeExporter.enabled to true" "" }}
{{- end }}
```

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/templates/grafana/dashboards-1.14/kubelet.yaml.patch:
```diff
@@ -9,3 +15,275 @@
+                         "targets": [
+                             {
+-                                "expr": "sum(up{cluster=\"$cluster\", job=\"kubelet\", metrics_path=\"/metrics\"})",
++                                "expr": "sum(up{cluster=\"$cluster\", job=\"{{ include "exporter.kubelet.jobName" . }}\", metrics_path=\"/metrics\"})",
```
metrics_path should come from the helper function throughout this file.

dbason replied:
Metrics path never changes here; this is the upstream value.  The only thing we're changing is the job name

aiyengar2 replied:
Ah sorry. I meant to delete this comment before submitting the review.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/templates/grafana/dashboards-1.14/controller-manager.yaml.patch:
```diff
@@ -20,7 +20,7 @@
                              {
 -                                "expr": "sum(up{job=\"kube-controller-manager\"})",
-+                                "expr": "sum(up{job=\"{{ include "exporter.kubeControllerManager.jobName" . }}\"})",
++                                "expr": "sum(up{job=\"{{ include "exporter.kubeControllerManager.jobName" . }}\"{{ include "exporter.controlplane.upLabels" . }}})",
```
directly patch this expression with an if/else if `k3sServer.enabled`.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/templates/grafana/dashboards-1.14/proxy.yaml.patch:
```diff
@@ -20,7 +20,7 @@
                              {
 -                                "expr": "sum(up{job=\"kube-proxy\"})",
-+                                "expr": "sum(up{job=\"{{ include "exporter.kubeProxy.jobName" . }}\"})",
++                                "expr": "sum(up{job=\"{{ include "exporter.kubeProxy.jobName" . }}\"{{ include "exporter.controlplane.upLabels" . }}})",
```
directly patch this with an if/else based on `k3sServer.enabled`.

aiyengar2 reviewed (changes requested):
Thanks for the changes! 

Overall looks good, but on a high level we should remove the `up` helper and move these changes directly into rancher-pushprox, since adding more functionality to customize the servicemonitor that is deployed would be universally useful, instead of maintaining a patch on the subchart.

dbason commented on packages/rancher-monitoring/generated-changes/patch/Chart.yaml.patch:
```diff
@@ -16,7 +16,16 @@
+   name: prometheus-adapter
+   repository: file://./charts/prometheus-adapter
+-- condition: nodeExporter.enabled
++- condition: nodeExporter.enabled,hardenedNodeExporter.enabled
```
Turns out it doesn't quite work like that:
```
Condition - The condition field holds one or more YAML paths (delimited by commas). If this path exists in the top parent's values and resolves to a boolean value, the chart will be enabled or disabled based on that boolean value. Only the first valid path found in the list is evaluated and if no paths exist then the condition has no effect.
```
If the first condition exists, and is falsely then the dependency won't be installed.

aiyengar2 replied:
@dbason thanks for confirming this. I wasn't aware that the first valid path would be taken, even if it was false.

Instead of using tags, would it work if we switch the order of conditions to `hardenedNodeExporter.enabled,nodeExporter.enabled` and removed `hardenedNodeExporter.enabled` from the values.yaml?

dbason replied:
That's what I was thinking of doing initially, but then hardenedNodeExporter would always be installed unless you explicitly set it to false:
> Only the first valid path found in the list is evaluated and if no paths exist then the condition has no effect.

aiyengar2 commented on packages/rancher-pushprox/charts/templates/pushprox-servicemonitor.yaml:
```diff
@@ -13,6 +13,37 @@ spec:
+    honorLabels: true
+    {{- end }}
+    {{- if .Values.endpoint.metricRelabelings }}
+    metricRelabelings:
+{{ tpl (toYaml .Values.endpoint.metricRelabelings | indent 4) . }}
```
```yaml
metricRelabelings: {{ tpl (toYaml .Values.endpoint.metricRelabelings) . | nindent 4 }}
```

aiyengar2 commented on packages/rancher-pushprox/charts/templates/pushprox-servicemonitor.yaml:
```diff
@@ -13,6 +13,37 @@ spec:
+{{ tpl (toYaml .Values.endpoint.metricRelabelings | indent 4) . }}
+    {{- end }}
+    {{- if .Values.endpoint.relabelings }}
+    relabelings:
+{{ toYaml .Values.endpoint.relabelings | indent 4 }}
```
use nindent

aiyengar2 commented on packages/rancher-pushprox/charts/templates/pushprox-servicemonitor.yaml:
```diff
@@ -13,6 +13,37 @@ spec:
+{{ toYaml .Values.endpoint.relabelings | indent 4 }}
+    {{- end }}
+  {{- range .Values.additionalEndpoints }}
+  - port: metrics
```
port should also be taken in from additionalEndpoints

aiyengar2 commented on packages/rancher-pushprox/charts/templates/pushprox-servicemonitor.yaml:
```diff
@@ -13,6 +13,37 @@ spec:
+    honorLabels: true
+    {{- end }}
+    {{- if .metricRelabelings }}
+    metricRelabelings:
+{{ tpl (toYaml .metricRelabelings | indent 4) . }}
```
use nindent

aiyengar2 commented on packages/rancher-pushprox/charts/templates/pushprox-servicemonitor.yaml:
```diff
@@ -13,6 +13,37 @@ spec:
+{{ tpl (toYaml .metricRelabelings | indent 4) . }}
+    {{- end }}
+    {{- if .relabelings }}
+    relabelings:
+{{ toYaml .relabelings | indent 4 }}
```
use nindent

aiyengar2 commented on packages/rancher-pushprox/charts/templates/pushprox-clients-rbac.yaml:
```diff
@@ -11,7 +11,12 @@ rules:
 {{- if and .Values.clients.https.enabled .Values.clients.https.useServiceAccountCredentials }}
-- nonResourceURLs: ["/metrics"]
+- nonResourceURLs:
+  - "/metrics"
+  - "/metrics/cadvisor"
```
`/metrics/cadvisor` should not be hard-coded; instead, it should come from the `path` provided under all `additionalEndpoints`.

aiyengar2 commented on packages/rancher-pushprox/charts/templates/pushprox-clients-rbac.yaml:
```diff
@@ -11,7 +11,12 @@ rules:
+  - "/metrics"
+  - "/metrics/cadvisor"
+  verbs: ["get"]
+- apiGroups: [""]
+  resources: ["nodes/metrics"]
```
Ideally, this should be supplied by the user and not hard-coded; not all ServiceMonitors need this permission.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/Chart.yaml.patch:
```diff
@@ -16,7 +16,26 @@
+ - condition: hardenedNodeExporter.enabled
+   name: hardenedNodeExporter
+   repository: file://./charts/hardenedNodeExporter
++  tags:
++  - hardenedNodeExporter
+ - condition: k3sServer.enabled
+   name: k3sServer
+   repository: file://./charts/k3sServer
+@@ -41,6 +52,9 @@
+ - condition: nodeExporter.enabled
+   name: prometheus-node-exporter
+   repository: file://./charts/prometheus-node-exporter
++  tags:
++  - nodeExporter
++  - hardenedNodeExporter
```
I'd like to avoid using tags since this would be a breaking change (for someone who chose to disable nodeExporter before). See https://github.com/rancher/charts/pull/1221#discussion_r643402266.

aiyengar2 commented on packages/rancher-monitoring/generated-changes/patch/templates/grafana/dashboards-1.14/controller-manager.yaml.patch:
```diff
@@ -20,7 +20,7 @@
                              {
 -                                "expr": "sum(up{job=\"kube-controller-manager\"})",
-+                                "expr": "sum(up{job=\"{{ include "exporter.kubeControllerManager.jobName" . }}\"})",
++                                "expr": "sum(up{job=\"{{ include "exporter.kubeControllerManager.jobName" . }}\"{{- if .Values.k3sServer.enabled  }}, metrics_path=\"/metrics\"{{- end }}})",
```
do you need the preceding `-` for the `if` and `end`?

aiyengar2 commented:
@dbason I forked your changes and made a couple of modifications based on the above comments I left behind.

Opened a new PR in https://github.com/rancher/charts/pull/1245; could you give it a review?

aiyengar2 commented:
Closing in favor of https://github.com/rancher/charts/pull/1245
